File: python/test_unilog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from unittest import main as unitMain, TestCase as unitTestCase
from unilog import UniLog, LogLevel


##########################################################################
# Logger basic verification

class TestUnilog(unitTestCase):
    
    ##########################################################################
    # Basic logging tests
    
    # Verify that primary logging methods work without crashing using devel profile
    def test_basic_logging(self):
        # We use 'standalone' and 'devel' for testing as they don't require external servers
        with UniLog(config_profile="standalone", app_name="test-basic", logger_profile="devel", log_level="debug") as logger:
            logger.info("Testing Python info log")
            logger.debug("Testing Python debug log")
            logger.warning("Testing Python warning log")
            logger.error("Testing Python error log")
            
            # Test configuration getting (standalone has some defaults)
            val = logger.get_config("logger", "level", "not_found")
            print(f"Logged level from config: {val}")

    # def test_config_updates(self):
    #     # This test is more complex due to the Go-Python callback deadlock potential
    #     # with the Python GIL and synchronous CGO calls.
    #     with UniLog(config_profile="standalone", app_name="test-config") as logger:
    #         updates = []
    #         def on_update(data):
    #             updates.append(data)
    #             print(f"Python received update: {data}")
    #
    #         logger.on_config_update(on_update)
    #         logger.set_config("test_section", "test_key", "test_value")
    #         time.sleep(0.5)
            
    def test_all_levels(self):
        """Verify that all log levels can be sent through the FFI boundary."""
        with UniLog(config_profile="standalone", app_name="test-levels", logger_profile="devel") as logger:
            for level in LogLevel:
                logger._log(level, f"Testing level: {level.name}")
            
            # Verify GetLevel
            logger.set_level(LogLevel.CRITICAL)
            self.assertEqual(logger.get_level(), LogLevel.CRITICAL)
            
            logger.set_level(LogLevel.DEBUG)
            self.assertEqual(logger.get_level(), LogLevel.DEBUG)

    def test_metadata(self):
        """Verify metadata management through the FFI boundary."""
        with UniLog(config_profile="standalone", app_name="test-metadata", logger_profile="devel") as logger:
            # Test individual metadata
            logger.add_metadata("test_key", "test_value")
            logger.info("Log with individual metadata")
            
            # Test bulk metadata
            meta = {
                "system": "antigravity",
                "parity": "hardened",
                "version": "1.2.0"
            }
            logger.set_metadata(meta)
            logger.info("Log with bulk metadata")


##########################################################################
# Entry point

if __name__ == "__main__":
    unitMain()

File: unilog/cpp/test_unilog.cpp
#include "UniversalLogger.hpp"
#include <iostream>
#include <cassert>
#include <string>

int main() {
    std::cout << ">>> Running C++ Unit Tests..." << std::endl;

    try {
        // 1. Test Instantiation
        UniLog logger("cpp-test-suite", "standalone", "devel", UniLog::DEBUG, false, 0);
        std::cout << "  - Instantiation passed" << std::endl;

        // 2. Test Get/Set Log Level
        assert(logger.get_level() == UniLog::DEBUG);
        logger.set_level(UniLog::WARNING);
        assert(logger.get_level() == UniLog::WARNING);
        logger.set_level(UniLog::DEBUG);
        std::cout << "  - Level Get/Set passed" << std::endl;

        // 3. Test Config Get/Set
        logger.set_config("test_section", "test_key", "value_123");
        std::string val = logger.get_config("test_section", "test_key", "default");
        assert(val == "value_123");

        std::string missing = logger.get_config("test_section", "non_existent_key", "default_val");
        assert(missing == "default_val");
        std::cout << "  - Config Get/Set passed" << std::endl;

        // 4. Test Metadata Manipulation
        logger.add_metadata("test_meta_key", "meta_val");
        logger.set_metadata("{\"test_env\":\"ci\"}");
        std::cout << "  - Metadata management passed" << std::endl;

        // 5. Test Macros & Logging (should execute without crash)
        UNILOG_INFO(logger, "C++ Info message from automated test");
        UNILOG_DEBUG(logger, "C++ Debug message from automated test");
        UNILOG_WARNING(logger, "C++ Warning message from automated test");
        UNILOG_ERROR(logger, "C++ Error message from automated test");
        UNILOG_CRITICAL(logger, "C++ Critical message from automated test");
        std::cout << "  - Logging macros execution passed" << std::endl;

        std::cout << ">>> ✅ ALL C++ UNIT TESTS PASSED SUCCESSFULLY!" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << ">>> ❌ C++ UNIT TEST FAILED: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
